### PR TITLE
Add events for player changing spectator target

### PR DESCRIPTION
**Pull request by CalebBassham**
- Add PlayerStartSpectatingEntityEvent
- Add PlayerStopSpectatingEntityEvent

This is my first pr here so if there is anything that should be different please let me know!

### CODE REVIEW

**Comment by mikroskeem** [comment]
Events could have some documentation.  
Do not modify imports.

Use Event#callEvent() instead of Bukkit.getPluginManager().callEvent(Event)

**Comment by CalebBassham** [comment]
I tried doing what you said and then updating the patches by following the steps of method 1 with the paper tool but after looking at the patches it doesn't look like any of my new changes are in them.

**Comment by CalebBassham** [comment]
I made the changes before the rebase and it kinda looks like intellij was just slow to update my file after the rebase reset it.

**Comment by CalebBassham** [comment]
I went back and redid the rebase and rebuilt the patches. It seems to have worked fine this time but there is a path that is not mine that was modified. This does not seem right to me so I wait to commit it.

here is the changes patch (updated on the right):
screenshot from 2018-09-28 04-16-37

**Comment by CalebBassham** [comment]
Is there anything else that I need to do here?

**Comment by CalebBassham** [comment]
As requested on Discord:
- Removed blank line from import statements
- Removed my events from import statements
- reverted the name of a variable to reduce diffs

File: Spigot-API-Patches/0157-Add-spectator-target-events.patch [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
++    private final Entity currentSpectatorTarget;
++    private final Entity newSpectatorTarget;
++
++    public PlayerStartSpectatingEntityEvent(Player player, Entity currentSpectatorTarget, Entity newSpectatorTarget) {
++        super(player);
++        this.currentSpectatorTarget = currentSpectatorTarget;
++        this.newSpectatorTarget = newSpectatorTarget;
++    }
++
++    public Entity getCurrentSpectatorTarget() {
```
Comment (zachbr) [inline Spigot-API-Patches/0157-Add-spectator-target-events.patch]: Even basic docs like "The entity being spectated" often lend clarity, please add them API side.

File: Spigot-API-Patches/0157-Add-spectator-target-events.patch [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
++        return currentSpectatorTarget;
++    }
++
++    public Entity getNewSpectatorTarget() {
```
Comment (zachbr) [inline Spigot-API-Patches/0157-Add-spectator-target-events.patch]: Docs

File: Spigot-API-Patches/0157-Add-spectator-target-events.patch [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
++    private static final HandlerList handlers = new HandlerList();
++    private boolean cancelled;
++    private final Entity spectatorTarget;
++
++    public PlayerStopSpectatingEntityEvent(Player player, Entity spectatorTarget) {
++        super(player);
++        this.spectatorTarget = spectatorTarget;
++    }
++
++    public Entity getSpectatorTarget() {
```
Comment (zachbr) [inline Spigot-API-Patches/0157-Add-spectator-target-events.patch]: Docs

File: Spigot-Server-Patches/0381-Call-player-spectator-target-events.patch [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+     }
+ 
+-    public void setSpectatorTarget(Entity entity) {
++    public void setSpectatorTarget(Entity newSpectatorTarget) {
+         Entity entity1 = this.getSpecatorTarget();
+ 
+-        this.cv = (Entity) (entity == null ? this : entity);
+-        if (entity1 != this.cv) {
++        if (newSpectatorTarget == null) {
++            this.cv = this;
```
Comment (zachbr) [inline Spigot-Server-Patches/0381-Call-player-spectator-target-events.patch]: all of these assignments to `this.cv` should probably use an obfhelper setter, with that setter up by the field declaration.
```java
private Entity cv; private void setSpectatorTarget(Entity e) { this.cv = e; } // Paper - OBFHELPER
```

Just so when we go to update this to 1.14 or 1.5512.4 or whatever its clear what the intent was even when this gets renamed.

**Review by zachbr — CHANGES_REQUESTED** [review]

**Comment by CalebBassham** [comment]
I will try and do that tomorrow.

File: Spigot-Server-Patches/0381-Call-player-spectator-target-events.patch (hunk 4 above)
Comment (CalebBassham) [reply Spigot-Server-Patches/0381-Call-player-spectator-target-events.patch]: `setSpectatorTarget(Entity)` is the method that I am changing for this pr. Should I do something like `setSpectatorTargetField(Entity)` as an obfhelper or keep it as it is?

Comment (zachbr) [reply Spigot-Server-Patches/0381-Call-player-spectator-target-events.patch]: Naming it slightly different is probably smart, I think that'd be fine 👍

**Comment by zachbr** [comment]
So some minor stuff I did
-  Merged the commits down into one patch, not one patch each. This is a pretty simple git rebase -i in each of the sub-repos
- Switched from casting to API interfaces to calling getBukkitEntity() on the entities
- Rebased all of these changes on top of master so we can merge them with little drama
- Super minor docs changes

Going to squash and merge, thanks for contributing. If you have any questions about what, why, or how, don't hesitate to ask.

**Review by zachbr — APPROVED** [review]

**Pull request merged**